hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,293 @@
+};
+
+#[cfg(feature = "min-samd51g")]
+pub use crate::target_device::dmac::channel::{
+    chctrla::{
+        BURSTLEN_A as BurstLength, THRESHOLD_A as FifoThreshold, TRIGACT_A as TriggerAction,
+        TRIGSRC_A as TriggerSource,
+    },
+    chprilvl::PRILVL_A as PriorityLevel,
+};
+
+use super::{
+    channel::{new_chan, Channel, Uninitialized},
+    DESCRIPTOR_SECTION, WRITEBACK,
+};
+use crate::target_device::{DMAC, PM};
+
+/// Trait representing a DMA channel ID
+pub trait ChId {
+    const U8: u8;
+    const USIZE: usize;
+}
+
+macro_rules! define_channels_struct {
+    ($num_channels:literal) => {
+        seq!(N in 0..$num_channels {
+            paste! {
+                #(
+                    /// Type alias for a channel number
+                    pub struct [<Ch N>];
+
+                    impl ChId for [<Ch N>] {
+                        const U8: u8 = N;
+                        const USIZE: usize = N;
+                    }
+                )*
+
+                /// Struct generating individual handles to each DMA channel
+                pub struct Channels(
+                    #(
+                        pub Channel<[<Ch N>], Uninitialized>,
+                    )*
+                );
+            }
+        });
+    };
+}
+
+with_num_channels!(define_channels_struct);
+
+/// Initialized DMA Controller
+pub struct DmaController {
+    dmac: DMAC,
+}
+
+#[bitfield]
+#[repr(u16)]
+pub struct PriorityLevelMask {
+    #[skip]
+    _reserved: B8,
+    /// Level 0
+    #[allow(dead_code)]
+    level0: bool,
+    /// Level 1
+    #[allow(dead_code)]
+    level1: bool,
+    /// Level 2
+    #[allow(dead_code)]
+    level2: bool,
+    /// Level 3
+    #[allow(dead_code)]
+    level3: bool,
+    #[skip]
+    _reserved: B4,
+}
+
+#[bitfield]
+#[repr(u32)]
+pub struct RoundRobinMask {
+    #[skip]
+    _reserved: B7,
+    /// Level 0
+    #[allow(dead_code)]
+    level0: bool,
+    #[skip]
+    _reserved: B7,
+    /// Level 1
+    #[allow(dead_code)]
+    level1: bool,
+    #[skip]
+    _reserved: B7,
+    /// Level 2
+    #[allow(dead_code)]
+    level2: bool,
+    #[skip]
+    _reserved: B7,
+    /// Level 3
+    #[allow(dead_code)]
+    level3: bool,
+}
+
+impl DmaController {
+    /// Return an immutable reference to the underlying DMAC object exposed by
+    /// the PAC.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because `DmaController` may expect certain
+    /// registers to retain a configuration. Messing with that configuration may
+    /// be unsafe.
+    pub unsafe fn dmac(&self) -> &DMAC {
+        &self.dmac
+    }
+
+    /// Return a mutable reference to the underlying DMAC object exposed by the
+    /// PAC.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because `DmaController` may expect certain
+    /// registers to retain a configuration. Messing with that configuration may
+    /// be unsafe.
+    pub unsafe fn dmac_mut(&mut self) -> &mut DMAC {
```

Comment:
I've started to question my own use of `&mut` references to PAC structs. There doesn't seem to be much value, since each PAC struct contains only `VolatileCell`s, which can be mutated through shared references. There doesn't seem to be any upside, but the downside is possible UB if you accidentally alias. Do you have any counter arguments? I haven't yet convinced myself one way or the other.

Reply:
Actually, isn't it instant UB if you have a simultaneous `&` reference and `&mut` reference? Seems risky. I'm leaning toward only using `&` from now on.

Reply:
I think you're actually right. I'm going with `pub(super) fn dmac(&mut self) -> &DMAC`.